test/common/http/http2/hpack_fuzz_test.cc

```diff
@@ -0,0 +1,147 @@
+  Buffer::OwnedImpl payload;
+  payload.reserve(buflen, &iovec, 1);
+  ASSERT(iovec.len_ >= buflen);
+
+  // Encode using nghttp2
+  uint8_t* buf = reinterpret_cast<uint8_t*>(iovec.mem_);
+  const ssize_t result =
+      nghttp2_hd_deflate_hd(deflater, buf, buflen, input_nv.begin(), input_nv.size());
+  if (result < 0) {
+    ENVOY_LOG_MISC(trace, "Failed to decode with result {}", result);
+    nghttp2_hd_deflate_del(deflater);
+    return absl::nullopt;
+  }
+
+  iovec.len_ = result;
+  payload.commit(&iovec, 1);
+
+  // Delete deflater.
+  nghttp2_hd_deflate_del(deflater);
+
+  return payload;
+}
+
+TestRequestHeaderMapImpl decodeHeaders(const Buffer::OwnedImpl& payload, bool end_headers) {
+  // Create inflater
+  nghttp2_hd_inflater* inflater;
+  const int rv = nghttp2_hd_inflate_new(&inflater);
+  ASSERT(rv == 0);
+
+  // Decode using nghttp2
+  Buffer::RawSliceVector slices = payload.getRawSlices();
+  const int num_slices = slices.size();
+  ASSERT(num_slices == 1, absl::StrCat("number of slices ", num_slices));
+
+  nghttp2_nv decoded_nv;
+  TestRequestHeaderMapImpl decoded_headers;
+  int inflate_flags = 0;
+  while (slices[0].len_ > 0) {
+    ssize_t result = nghttp2_hd_inflate_hd2(inflater, &decoded_nv, &inflate_flags,
+                                            reinterpret_cast<uint8_t*>(slices[0].mem_),
+                                            slices[0].len_, end_headers);
+    // Decoding should not fail and data should not be left in slice.
+    ASSERT(result >= 0 && (slices[0].len_ = 0));
+
+    slices[0].mem_ = reinterpret_cast<void*>(reinterpret_cast<uint8_t*>(slices[0].mem_) + result);
+    slices[0].len_ -= result;
+
+    if (inflate_flags & NGHTTP2_HD_INFLATE_EMIT) {
+      // One header key value pair has been successfully decoded.
+      decoded_headers.addCopy(
+          std::string(reinterpret_cast<char*>(decoded_nv.name), decoded_nv.namelen),
+          std::string(reinterpret_cast<char*>(decoded_nv.value), decoded_nv.valuelen));
+    }
+  }
+
+  if (end_headers) {
+    nghttp2_hd_inflate_end_headers(inflater);
+  }
+
+  // Delete inflater
+  nghttp2_hd_inflate_del(inflater);
+
+  return decoded_headers;
+}
+
+DEFINE_PROTO_FUZZER(const test::common::http::http2::HpackTestCase& input) {
+  // Validate headers.
+  try {
+    TestUtility::validate(input);
+  } catch (const EnvoyException& e) {
+    ENVOY_LOG_MISC(trace, "EnvoyException: {}", e.what());
+    return;
+  }
+
+  // Create name value pairs from headers.
+  const TestRequestHeaderMapImpl headers =
+      Fuzz::fromHeaders<TestRequestHeaderMapImpl>(input.headers());
+  const absl::FixedArray<nghttp2_nv> input_nv = createNameValueArray(headers);
+
+  // Encode headers with nghttp2.
+  ENVOY_LOG_MISC(trace, "Encoding headers {}", headers);
+  const absl::optional<Buffer::OwnedImpl> payload = encodeHeaders(input_nv);
+  if (!payload.has_value() || payload.value().getRawSlices().size() == 0) {
+    // An empty header map produces no payload, skip decoding.
```

Comment:
What if the empty payload is due to some nghttp2 failure?

Reply:
If nghttp2 throws a negative error code, payload will be a nullopt, which will return. I think this boils down to whether the team wants scope of fuzzer to be to simply check if encode/decode always work by sending headers through both and checking for equality, or if the team wants to explore header values that could cause nghttp2 to throw errors.

Reply:
I think catching unexpected errors from nghttp2 would be a nice bonus, since we have a lot of ASSERTs in places around the returns from nghttp2 calls.

Reply:
Done. Empty payload is now handled earlier, since in our codebase and here we pass `.data()`, which, if nullptr because of empty header map, will crash nghttp2 at runtime. That's kinda bad, but I went through the codebase and there's nowhere we supply headers via `data()` without at least one header. We should add a statement enforcing this though, maybe.

Since empty payload is not worrisome, I catch errors with an ASSERT(result >= 0) now.